src/include/mpir_comm.h

```diff
@@ -344,20 +344,28 @@ MPL_STATIC_INLINE_PREFIX int MPIR_Stream_comm_set_attr(MPIR_Comm * comm, int src
 
 MPL_STATIC_INLINE_PREFIX int MPIR_Get_internode_rank(MPIR_Comm * comm, int r)
 {
-    /* TODO: optimize for special ones such as node_consecutive and balanced */
     MPIR_Assert(comm->attr | MPIR_COMM_ATTR__HIERARCHY);
-    MPIR_Assert(comm->internode_table);
-
-    return comm->internode_table[r];
+    if (comm->internode_table) {
+        return comm->internode_table[r];
+    } else {
+        /* canonical or trivial */
+        return r / comm->num_local;
+    }
 }
 
 MPL_STATIC_INLINE_PREFIX int MPIR_Get_intranode_rank(MPIR_Comm * comm, int r)
 {
-    /* TODO: optimize for special ones such as node_consecutive and balanced */
     MPIR_Assert(comm->attr | MPIR_COMM_ATTR__HIERARCHY);
-    MPIR_Assert(comm->intranode_table);
-
-    return comm->intranode_table[r];
+    if (comm->intranode_table) {
+        return comm->intranode_table[r];
+    } else {
+        /* canonical or trivial */
+        if ((comm->rank / comm->num_local) == (r / comm->num_local)) {
+            return r % comm->num_local;
```

Comment:
is there another minor optimization to be had in the node consecutive case where we can subtract and offset rather than use the modulo operator? otherwise this looks good.

Reply:
Yes, but you still need maintain an array of node offsets. I don't think the use case will be common enough for the extra code.

Reply:
I'm only thinking of the node consecutive case and where we've already determined they are on the same node. Something like:
```c
/* canonical or trivial */
if ((comm->rank / comm->num_local) == (r / comm->num_local)) {
    if (comm->hierarchy_flags & MPIR_COMM_HIERARCHY__NODE_CONSECUTIVE) {
        return r - comm->local_offset;
    }
    return r % comm->num_local;
} else {
    return -1;
}
```
But we're already 3 branches deep so maybe it's not worth it.

Reply:
You can't reliably use `comm->rank / comm->num_local` when `comm->num_local` may not represent node boundary.

Reply:
I'm suggesting a not-yet-defined `comm->local_offset` or `comm->node_offset` which you could compute at creation time. Anyway we don't need to hold up this PR on it.